### PR TITLE
Fix test flakiness

### LINKED ISSUE #16986: Flaky test TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset
https://github.com/trinodb/trino/actions/runs/4675961675/jobs/8282707441 (https://github.com/trinodb/trino/pull/16970)

```
2023-04-12T03:43:56.067-0600	ERROR	pool-24-thread-2	io.trino.testng.services.ProgressLoggingListener	[TEST FAILURE] io.trino.execution.scheduler.TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset; (took: 0.0 seconds)
java.lang.AssertionError: node lease acquired expected [false] but found [true]
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:513)
	at org.testng.Assert.assertFalse(Assert.java:63)
	at io.trino.execution.scheduler.TestBinPackingNodeAllocator.assertNotAcquired(TestBinPackingNodeAllocator.java:678)
	at io.trino.execution.scheduler.TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset(TestBinPackingNodeAllocator.java:385)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset:385->assertNotAcquired:678 node lease acquired expected [false] but found [true]
[INFO] 
Error:  Tests run: 7996, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
```

### PR DESCRIPTION
**Pull request by losipiuk**
## Description

fixes #16986

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java [hunk 1]
```diff
@@ -73,7 +73,7 @@ public Set<InternalNode> getNodes(NodeState state)
     @Override
     public Set<InternalNode> getActiveCatalogNodes(CatalogHandle catalogHandle)
     {
-        return allNodes;
+        return ImmutableSet.copyOf(allNodes);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java:76]: same in `public Set<InternalNode> getNodes(NodeState state)`

also same when constructing `return new NodesSnapshot(` / `return new AllNodes(` -- these are separate classes defined in a different file, we shouldn't assume they do defensive copy in their constructors.

File: core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java [hunk 2]
```diff
@@ -68,5 +69,14 @@ public Set<InternalNode> getConnectorNodes(CatalogHandle catalogHandle)
                     .map(map -> map.get(catalogHandle))
                     .orElse(allNodes);
         }
+
+        @Override
+        public String toString()
+        {
+            return MoreObjects.toStringHelper(this)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java:76]: import sstatically

File: core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java [hunk 3]
```diff
@@ -68,5 +69,14 @@ public Set<InternalNode> getConnectorNodes(CatalogHandle catalogHandle)
...
+                    .add("allNodes", allNodes)
+                    .add("connectorNodes", connectorNodes)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java:78]: consider `.orElse(null)` for shorter tostring

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java [hunk 4]
```diff
@@ -380,6 +381,10 @@ public void testNoMatchingNodeAvailableTimeoutReset()
 
             // remove the node - we are again in situation where no matching nodes exist in cluster
             nodeManager.removeNode(NODE_2);
+
+            // sleep for a while before releasing lease, as background processPendingAcquires may be still be running with old snapshot
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java:385]: stuttered "be"

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java [hunk 5]
```diff
@@ -380,6 +381,10 @@ public void testNoMatchingNodeAvailableTimeoutReset()
...
+            // containing NODE_2, and theNotAcquireLease could be fulfilled when theAcquireLease is released
+            sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java:387]: can this be synchronized without a sleep?

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java (hunk 5 above)
Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java:387]: Not really - without changing how we setup the test object (that is disabling it's internal thread) - which I think is worse solution as we end up testing something less resembling what we are gonna use in production.
I know it is not optimal but I think it should do the job.

**Review by linzebing — APPROVED** [review]

**Pull request merged**